core/storage/trie_pruner/impl/trie_pruner_impl.cpp

```diff
@@ -0,0 +1,466 @@
+                                version,
+                                AddConfig{
+                                    .type = AddConfig::AddLoadedOnly,
+                                }));
+    OUTCOME_TRY(savePersistentState());
+    return outcome::success();
+  }
+
+  outcome::result<void> TriePrunerImpl::addNewChildState(
+      storage::trie::RootHash const &parent_root,
+      trie::PolkadotTrie const &new_trie,
+      trie::StateVersion version) {
+    OUTCOME_TRY(child_root_hash,
+                addNewStateWith(new_trie,
+                                version,
+                                AddConfig{
+                                    .type = AddConfig::AddLoadedOnly,
+                                }));
+    OUTCOME_TRY(markAsChild(Parent{parent_root}, Child{child_root_hash}));
+    return outcome::success();
+  }
+
+  outcome::result<void> TriePrunerImpl::markAsChild(Parent parent,
+                                                    Child child) {
+    child_states_[parent.hash].emplace_back(child.hash);
+    OUTCOME_TRY(savePersistentState());
+    return outcome::success();
+  }
+
+  outcome::result<storage::trie::RootHash> TriePrunerImpl::addNewStateWith(
+      trie::PolkadotTrie const &new_trie,
+      trie::StateVersion version,
+      AddConfig config) {
+    if (new_trie.getRoot() == nullptr) {
+      SL_DEBUG(logger_, "Attempt to add a trie with a null root");
+      return outcome::success();
+    }
+
+    SL_DEBUG(logger_, "Ref count map size is {}", ref_count_.size());
+    KAGOME_PROFILE_START_L(logger_, register_state);
+
+    std::vector<std::shared_ptr<const trie::OpaqueTrieNode>> queued_nodes;
+    queued_nodes.push_back(new_trie.getRoot());
+
+    EncoderCache encoder{*codec_, logger_};
+
+    OUTCOME_TRY(root_value,
+                encoder.getMerkleValue(*new_trie.getRoot(), version));
+    OUTCOME_TRY(root_hash, calcHash(*codec_, *new_trie.getRoot(), version));
+
+    ref_count_[root_value] += 1;
+    logger_->info("Add: Merkle value of state {} is {}", root_hash, root_value);
+
+    size_t referenced_nodes_num = 0;
+
+    while (!queued_nodes.empty()) {
+      auto opaque_node = queued_nodes.back();
+      queued_nodes.pop_back();
+      OUTCOME_TRY(hash, encoder.getMerkleValue(*opaque_node, version));
+      SL_TRACE(logger_,
+               "Add - Node {}, ref count {}",
+               hash.toHex(),
+               ref_count_[hash]);
+
+      referenced_nodes_num++;
+      if (auto node = dynamic_cast<trie::TrieNode const *>(opaque_node.get());
+          node != nullptr && node->isBranch()
+          && (config.shouldAddAllNodes() || ref_count_[hash] == 1)) {
+        auto branch = static_cast<const trie::BranchNode *>(opaque_node.get());
+        for (auto opaque_child : branch->children) {
+          if (opaque_child != nullptr) {
+            OUTCOME_TRY(child_hash,
+                        encoder.getMerkleValue(*opaque_child, version));
+            ref_count_[child_hash] += 1;
```

Comment:
Increment ref for "queue" item, not for "child".
Otherwise `ref_count_[hash] == 1` may be false and miss children.
```suggestion
```